### ISSUE TITLE
VAT ID research

### ISSUE THREAD

**Issue by masonmcelvain** [iFixit/checkout-ui-extensions#10]
## Problem

We want to support VAT ID validation and discount application during checkout.

On shopify-v2, we use the eu vat validator app for this.

## Do this

Research whether we can still use the vat validator app from the checkout extension app. If so, open a pull. If not, make an issue for an alternative solution.

**Comment by masonmcelvain** [comment]
It doesn't seem like we have access to the checkout ID in checkout ui extensions, so I don't think we can use the EU VAT validator to mutate the checkout. We don't even have access to the checkout url, since the code runs in a web worker ([ref](https://shopify.dev/docs/api/checkout-ui-extensions/unstable/configuration#network-access)).

One way to implement VAT check:
- Checkout UI extensions renders the VAT field
- User types their VAT and submits it
- UI extension makes a network request to an iFixit server that then talks to VIES or evatr to validate the VAT ID ([network access docs](https://shopify.dev/docs/api/checkout-ui-extensions/unstable/configuration#network-access))
  - We have access to checkout line items ([ref](https://shopify.dev/docs/api/checkout-ui-extensions/unstable/react-hooks/cart/usecartlines)), and can also request access to the customer shipping address company and city ([ref](https://shopify.dev/docs/api/checkout-ui-extensions/unstable/react-hooks/buyer-identity/useshippingaddress)).
- If the VAT ID is valid, then the iFixit server can [create](https://shopify.dev/docs/api/admin-graphql/2023-04/mutations/discountCodeBasicCreate) a discount code via the Admin API and send it back to the UI extension
  - The code can only work for that customer, and can be used only once, and can even have an expiration if we want
- The UI extension [applies](https://shopify.dev/docs/api/checkout-ui-extensions/unstable/react-hooks/reductions/useapplydiscountcodechange) the discount code to checkout

With this approach, we would need to be careful not to let the user have more than one VAT discount code.

**Comment by danielbeardsley** [comment]
@masonmcelvain and I talked for a while, researched, and came up with a new plan :point_up:
